net: fuzz: bypass network magic and checksum validation

brunoerg opened the pull request:
Fixes #30957

We currently have some instructions on the documentation about how to fuzz the Bitcoin Core P2P layer using Honggfuzz NetDriver. It includes a patch to be applied to bypass network magic and checksum validation, that is outdated. However, we can change the code to bypass them using `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION`. It also makes the `p2p_transport_serialization` fuzz target simpler since we do not need to assist the network magic and checksum creation anymore.

brunoerg committed:
net: bypass network magic and checksum validation for fuzzing

brunoerg committed:
doc: fuzz: remove instructions to bypass network magic and checksum validation for Honggfuzz NetDriver

brunoerg committed:
fuzz: remove checksum and magic bytes assist in `p2p_transport_serialization`

brunoerg commented:
cc: @dergoegge

dergoegge commented:
Concept ACK

dergoegge commented on doc/fuzzing.md:
```diff
@@ -241,31 +241,6 @@ index 8195bceaec..cce2b31ff0 100644
-                  SanitizeString(msg.m_type), msg.m_message_size,
-                  HexStr(Span{hash}.first(CMessageHeader::CHECKSUM_SIZE)),
 EOF
 $ cmake -B build_fuzz \
```
`CPPFLAGS` will have to include `-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` when building bitcoind for fuzzing with the honggfuzz netdriver.

brunoerg replied:
Isn't `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` set when using `-DBUILD_FOR_FUZZING=ON`?

```
if(BUILD_FOR_FUZZING)
  message(WARNING "BUILD_FOR_FUZZING=ON will disable all other targets and force BUILD_FUZZ_BINARY=ON.")
  set(BUILD_DAEMON OFF)
  set(BUILD_CLI OFF)
  set(BUILD_TX OFF)
  set(BUILD_UTIL OFF)
  set(BUILD_UTIL_CHAINSTATE OFF)
  set(BUILD_KERNEL_LIB OFF)
  set(BUILD_WALLET_TOOL OFF)
  set(BUILD_GUI OFF)
  set(ENABLE_EXTERNAL_SIGNER OFF)
  set(WITH_MINIUPNPC OFF)
  set(WITH_ZMQ OFF)
  set(BUILD_TESTS OFF)
  set(BUILD_GUI_TESTS OFF)
  set(BUILD_BENCH OFF)
  set(BUILD_FUZZ_BINARY ON)

  target_compile_definitions(core_interface INTERFACE
    ABORT_ON_FAILED_ASSUME
    FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
  )
endif()
```

brunoerg replied:
Yes, missed that. Just added it.

brunoerg commented:
Just moved it to draft to get some conceptual feedbacks first. I added a commit to bypass some asserts in `ConnmanTestMsg` that would fail since we're bypassing network magic and checksum validation. Also, I think we will have to change `p2p_transport_bidirectional` target since there are some sanity checks that would fail with the bypasses. Any thoughts?

dergoegge commented:
> They are unrelated to the checksum, so if they are failing, it seems like this patch is incomplete and some part is still checking the magic or checksum, no?

I'm guessing that we need to make sure that (if we are fuzzing) `V1Transport::SetMessageToSend` creates valid checksums/magic-bytes according to the new fuzzing mode check.

brunoerg commented:
> Can you explain why the asserts and sanity checks would be failing? They are unrelated to the checksum, so if they are failing, it seems like this patch is incomplete and some part is still checking the magic or checksum, no?

Some previously valid magic or checksum could be now be considered invalid. But I removed the commit that bypasses the asserts and changed the verification to: 
```
#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
    if ((hdr.pchMessageStart[0] & 1) != 0 && hdr.pchMessageStart != m_magic_bytes) {
#else
```
instead of just:
```
#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
    if ((hdr.pchMessageStart[0] & 1) != 0) {
#else
```

I think this way we still facilitate it without invalidating valid magic/checksum.

brunoerg commented on src/net.cpp:
```diff
@@ -728,7 +728,11 @@ int V1Transport::readHeader(Span<const uint8_t> msg_bytes)
     }
 
     // Check start string, network magic
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+    if ((hdr.pchMessageStart[0] & 1) != 0 && hdr.pchMessageStart != m_magic_bytes) {
+#else
     if (hdr.pchMessageStart != m_magic_bytes) {
+#endif
```
Before, if `hdr.pchMessageStart != m_magic_bytes` we would return an error. That is, fuzzing would have to spend time matching it. Now, we will only invalidate it if `hdr.pchMessageStart != m_magic_bytes` and `(hdr.pchMessageStart[0] & 1) != 0`. So, it will make fuzzing easier by just needing to match `(hdr.pchMessageStart[0] & 1) == 0`.

dergoegge commented on doc/fuzzing.md:
```diff
@@ -198,6 +198,7 @@ $ cmake -B build_fuzz \
    -DCMAKE_C_COMPILER="$(pwd)/honggfuzz/hfuzz_cc/hfuzz-clang" \
    -DCMAKE_CXX_COMPILER="$(pwd)/honggfuzz/hfuzz_cc/hfuzz-clang++" \
    -DBUILD_FOR_FUZZING=ON \
+   -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=ON \
```
This won't define the c++ macro `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION`, as this only sets a cmake build flag (cmake probably complains that this flag doesn't exist).

You should be able to use `-DAPPEND_CPPFLAGS="-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION"` (I see where the confusion comes from, as `-D` is accepted by compilers to define macros and by cmake to set build flags but they are different things). 

You will also need to add this to the correct section in the docs (i.e. the netdriver docs):

https://github.com/bitcoin/bitcoin/blob/ffdbf6c45be2a69050401cd5dad785758cee21d4/doc/fuzzing.md?plain=1#L246-L258

brunoerg replied:
> You should be able to use -DAPPEND_CPPFLAGS="-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION" (I see where the confusion comes from, as -D is accepted by compilers to define macros and by cmake to set build flags but they are different things).

Ah yes, thank you. Just saw some examples from the CI scripts.

brunoerg commented on src/net.cpp:
```diff
@@ -793,7 +797,11 @@ CNetMessage V1Transport::GetReceivedMessage(const std::chrono::microseconds time
     RandAddEvent(ReadLE32(hash.begin()));
 
     // Check checksum and header message type string
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+    if ((hdr.pchChecksum[0] & 1) != 0 && memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0) {
+#else
     if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0) {
+#endif
```
Nice, I will address it.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31012#discussion_r1785875497 and https://github.com/bitcoin/bitcoin/pull/31012#discussion_r1785885817. Thanks, @dergoegge.

marcofleon commented:
Approach ACK. The use of `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` in `net.cpp` looks good to me. Generated a coverage report to verify that both branches of both checks are hit.

I haven't yet worked through the honggfuzz instructions to confirm that they work. I'll circle back to this to give it a try.

dergoegge commented:
Copy and pasting the command from the docs doesn't work:

```
$ git apply << "EOF" ...
error: corrupt patch at line 15
```

brunoerg commented:
> Copy and pasting the command from the docs doesn't work:

Just fixed it. Can you check it?

dergoegge commented:
`-v2transport=0` should be added to the bitcoind args used for netdriver fuzzing. Otherwise, it doesn't seem like honggfuzz ever fuzzes anything besides the v2 handshake.

Also playing around with this and reading the [netdriver post](https://blog.swiecki.net/2018/01/fuzzing-tcp-servers.html) (i couldn't find any other docs on it) I'm not sure if this will ever even get past the version handshake? So I'm wondering how useful this tool really is for us.

brunoerg commented:
> -v2transport=0 should be added to the bitcoind args used for netdriver fuzzing. Otherwise, it doesn't seem like honggfuzz ever fuzzes anything besides the v2 handshake.

Well noticed, just added it.

brunoerg commented:
> Also playing around with this and reading the [netdriver post](https://blog.swiecki.net/2018/01/fuzzing-tcp-servers.html) (i couldn't find any other docs on it) I'm not sure if this will ever even get past the version handshake? So I'm wondering how useful this tool really is for us.

Yes, I don't think it will pass the version handshake. I can't see what kind of bugs it would find that other fuzzers would not get.

dergoegge reviewed (approved):
ACK d43966603bdefb0402a2dc93ff990cb254cdf2df

I think we could consider removing the honggfuzz netdriver instructions in a follow up but the changes look good to me anyway. I tested this by running honggfuzz in netdriver mode.

marcofleon reviewed (commented):
Tested ACK d43966603bdefb0402a2dc93ff990cb254cdf2df

theuni commented:
If we're going to be peppering the code with these, could we add a safety harness to make sure it's never being used accidentally for a live node?

Maybe somewhere in init.cpp or bitcoind.cpp:
```c++
#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
exit(1);
#endif
```

brunoerg commented:
> If we're going to be peppering the code with these, could we add a safety harness to make sure it's never being used accidentally for a live node?

Seems good. Note that we're already use `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` to bypass `CheckProofOfWork`.

fanquake commented:
> could we add a safety harness to make sure it's never being used accidentally for a live node?
> Maybe somewhere in init.cpp or bitcoind.cpp:
> Edit: I realize that bitcoind should be disabled in that case. This would be belt-and-suspenders to make sure that always holds.

Just want to note that putting something in either of these files wouldn't be enough to prevent someone producing a kernel lib with the fuzzing code.

brunoerg commented:
> Can you provide more details here? IIUC a hash will still be calculated, just not compared, so this may only drop one hash call, with a possible maximum speed-up of 2x. However, it would be good to actually measure the difference. Otherwise, I am not sure if this is worth it.

This is not about performance. It makes the `p2p_transport_serialization` fuzz target simpler because we can remove the code that assists the checksum and magic bytes creation. So, we can achieve the same with less LoC.

brunoerg commented:
Since everyone agrees on removing Honggfuzz netdriver, I think we can change the approach to simply just remove it from the documentation.

brunoerg commented:
Closing this in favor of #31092
